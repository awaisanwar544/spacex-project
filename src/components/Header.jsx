import React from 'react';
import { NavLink } from 'react-router-dom';
import logo from '../assets/planet.png';

const Header = () => (
  <div id="navbar">
    <div>
      <img id="logo" src={logo} alt="logo" />
    </div>
    <nav>
      <li><NavLink to="/" activeclassname="active">Rockets</NavLink></li>
      <li><NavLink to="/missions" activeclassname="active">Missions</NavLink></li>
      <li><NavLink to="/profile" activeclassname="active">Profile</NavLink></li>
    </nav>
  </div>
);

export default Header;
