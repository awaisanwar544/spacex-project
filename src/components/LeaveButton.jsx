import React from 'react';
import { PropTypes } from 'prop-types';
import styled from 'styled-components';

const Wrapper = styled.div`
  button {
    background-color: #fff;
    border: 0.1em solid #9a9a9a;
    color: #9a9a9a;
    padding: 15px 32px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 16px;
    cursor: pointer;
    border-radius: 0.5em;
  }
`;

const LeaveButton = ({ id }) => (

  <Wrapper>
    <button id={id} type="button">Cancel Reservation</button>
  </Wrapper>
);

LeaveButton.propTypes = {
  id: PropTypes.number.isRequired,
};

export default LeaveButton;
