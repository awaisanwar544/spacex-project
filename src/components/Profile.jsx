import React from 'react';
import { useSelector } from 'react-redux';
import styled from 'styled-components';

const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  text-align: left;
  column-gap: 2em;
  padding: 1em 2em;

  div {
    width: 50%;
  }

  table {
    width: 100%;
  }

  td {
    padding: 1em;
  }

  tr {
    outline: 1px solid black;
  }
`;

const Profile = () => {
  const rockets = useSelector((state) => state.rockets);
  const booked = rockets.filter((item) => {
    if (item.reserved === true) {
      return item;
    }
    return '';
  });
  return (
    <Wrapper>
      <div>
        <h2>My Missions</h2>
        <table>
          <tbody>
            {booked.map((item) => (
              <tr key={item.id}>
                <td>{item.rocket_name}</td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
      <div>
        <h2>My Rockets</h2>
        <table>
          <tbody>
            {booked.map((item) => (
              <tr key={item.id}>
                <td>{item.rocket_name}</td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </Wrapper>
  );
};

export default Profile;
