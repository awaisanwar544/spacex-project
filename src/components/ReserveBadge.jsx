import React from 'react';
import styled from 'styled-components';

const Wrapper = styled.span`
    background-color: green;
    border: none;
    color: white;
    padding: 0.25em;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 0.75em;
    border-radius: 0.5em;
    margin-right: 1em;
`;

const ReserveBadge = () => (

  <Wrapper>
    Reserved
  </Wrapper>
);

export default ReserveBadge;
