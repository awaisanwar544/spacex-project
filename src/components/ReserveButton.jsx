import React from 'react';
import { PropTypes } from 'prop-types';
import styled from 'styled-components';

const Wrapper = styled.div`
  button {
    background-color: #1c2cdb;
    border: none;
    color: white;
    padding: 15px 32px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 16px;
    cursor: pointer;
    border-radius: 0.5em;
  }
`;

const ReserveButton = ({ id }) => (

  <Wrapper>
    <button id={id} type="button">Reserve Rocket</button>
  </Wrapper>
);

ReserveButton.propTypes = {
  id: PropTypes.number.isRequired,
};

export default ReserveButton;
