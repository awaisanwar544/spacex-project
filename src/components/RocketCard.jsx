import React from 'react';
import { PropTypes } from 'prop-types';
import styled from 'styled-components';
import ReserveButton from './ReserveButton';
import LeaveButton from './LeaveButton';
import ReserveBadge from './ReserveBadge';

const Wrapper = styled.div`
  display: flex;
  column-gap: 1em;
  padding: 1em 2em;

  img {
    width: 100%;
  }

  h3 {
    margin-top: 0;
  }

  p {
    margin-top: 0;
  }
`;

const ImgDiv = styled.div`
  flex-grow: 1;
  max-width: 15em;
`;

const ContentDiv = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  flex-grow: 3;
`;

const RocketCard = (props) => {
  const {
    id, rocketName, description, flickrImages, reserved,
  } = props;
  return (
    <Wrapper>
      <ImgDiv>
        <img src={flickrImages} alt="" />
      </ImgDiv>
      <ContentDiv>
        <h3>{rocketName}</h3>
        <p>
          <span>
            {reserved && (
              <ReserveBadge />
            )}
          </span>
          {description}
        </p>
        {reserved && (
        <LeaveButton id={id} />
        )}
        {!reserved && (
        <ReserveButton id={id} />
        )}
      </ContentDiv>
    </Wrapper>
  );
};

RocketCard.propTypes = {
  id: PropTypes.number.isRequired,
  rocketName: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  flickrImages: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};

export default RocketCard;
