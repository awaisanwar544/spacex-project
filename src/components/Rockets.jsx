import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets } from '../redux/rockets/rockets';
import RocketCard from './RocketCard';

const Rockets = () => {
  const rockets = useSelector((state) => state.rockets);
  const dispatch = useDispatch();
  useEffect(() => {
    if (rockets.length === 0) {
      dispatch(fetchRockets());
    }
  }, []);
  return (
    <div>
      {rockets.map((item) => (
        <RocketCard
          key={item.id}
          id={item.id}
          rocketName={item.rocket_name}
          description={item.description}
          flickrImages={item.flickr_images[0]}
          reserved={item.reserved}
        />
      ))}
    </div>
  );
};

RocketCard.defaultProps = {
  reserved: false,
};

export default Rockets;
