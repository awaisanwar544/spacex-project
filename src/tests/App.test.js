import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import store from '../redux/configureStore';
import Rockets from '../components/Rockets';
import Profile from '../components/Profile';
import Header from '../components/Header';
import RocketCard from '../components/RocketCard';
import ReserveButton from '../components/ReserveButton';
import LeaveButton from '../components/LeaveButton';
import ReserveBadge from '../components/ReserveBadge';

jest.mock('../components/ReserveButton');
jest.mock('../components/LeaveButton');

describe('App Components Render Correctly', () => {
  it('Header renders correctly', () => {
    const tree = renderer.create(
      <BrowserRouter>
        <Header />
      </BrowserRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Rockets renders correctly with empty store', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Rocket card renders correctly when reserved is false', () => {
    const tree = renderer.create(
      <RocketCard id={1} rocketName="" description="" flickrImages="" reserved={false} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('ReserveButton renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <ReserveButton id={1} />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('LeaveButton renders correctly', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <LeaveButton id={1} />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Rocket card renders correctly when reserved is true', () => {
    const tree = renderer.create(
      <RocketCard id={1} rocketName="" description="" flickrImages="" reserved />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('ReserveBadge renders correctly', () => {
    const tree = renderer.create(
      <ReserveBadge />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Profile renders correctly with empty store', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <Profile />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
