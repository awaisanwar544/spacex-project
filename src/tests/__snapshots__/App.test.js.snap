// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Components Render Correctly Header renders correctly 1`] = `
<div
  id="navbar"
>
  <div>
    <img
      alt="logo"
      id="logo"
      src="planet.png"
    />
  </div>
  <nav>
    <li>
      <a
        activeclassname="active"
        aria-current="page"
        className="active"
        href="/"
        onClick={[Function]}
      >
        Rockets
      </a>
    </li>
    <li>
      <a
        activeclassname="active"
        className=""
        href="/missions"
        onClick={[Function]}
      >
        Missions
      </a>
    </li>
    <li>
      <a
        activeclassname="active"
        className=""
        href="/profile"
        onClick={[Function]}
      >
        Profile
      </a>
    </li>
  </nav>
</div>
`;

exports[`App Components Render Correctly LeaveButton renders correctly 1`] = `null`;

exports[`App Components Render Correctly Profile renders correctly with empty store 1`] = `
<div
  className="sc-eCYdqJ vjkub"
>
  <div>
    <h2>
      My Missions
    </h2>
    <table>
      <tbody />
    </table>
  </div>
  <div>
    <h2>
      My Rockets
    </h2>
    <table>
      <tbody />
    </table>
  </div>
</div>
`;

exports[`App Components Render Correctly ReserveBadge renders correctly 1`] = `
<span
  className="sc-bczRLJ UMZIj"
>
  Reserved
</span>
`;

exports[`App Components Render Correctly ReserveButton renders correctly 1`] = `null`;

exports[`App Components Render Correctly Rocket card renders correctly when reserved is false 1`] = `
<div
  className="sc-gsnTZi fBTMkP"
>
  <div
    className="sc-dkzDqf hkBpzE"
  >
    <img
      alt=""
      src=""
    />
  </div>
  <div
    className="sc-hKMtZM kgqlvy"
  >
    <h3 />
    <p>
      <span />
    </p>
  </div>
</div>
`;

exports[`App Components Render Correctly Rocket card renders correctly when reserved is true 1`] = `
<div
  className="sc-gsnTZi fBTMkP"
>
  <div
    className="sc-dkzDqf hkBpzE"
  >
    <img
      alt=""
      src=""
    />
  </div>
  <div
    className="sc-hKMtZM kgqlvy"
  >
    <h3 />
    <p>
      <span>
        <span
          className="sc-bczRLJ UMZIj"
        >
          Reserved
        </span>
      </span>
    </p>
  </div>
</div>
`;

exports[`App Components Render Correctly Rockets renders correctly with empty store 1`] = `<div />`;
